kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt

```diff
@@ -0,0 +1,760 @@
+            .joinToString(", ") { "\"$it\"" }})",
+            optionsCN)
+        .build()
+
+    val adapter = TypeSpec.classBuilder(adapterName)
+        .superclass(jsonAdapterTypeName)
+        .apply {
+          genericTypeNames?.let {
+            addTypeVariables(genericTypeNames)
+          }
+        }
+        .apply {
+          // TODO make this configurable. Right now it just matches the source model
+          if (visibility == INTERNAL) {
+            addModifiers(KModifier.INTERNAL)
+          }
+        }
+        .primaryConstructor(FunSpec.constructorBuilder()
+            .addParameter(moshiParam)
+            .apply {
+              genericTypeNames?.let {
+                addParameter(typesParam)
+              }
+            }
+            .build())
+        .addProperty(optionsProperty)
+        .addProperties(adapterProperties.values)
+        .addFunction(FunSpec.builder("toString")
+            .addModifiers(OVERRIDE)
+            .returns(String::class)
+            .addStatement("return %S",
+                "GeneratedJsonAdapter(${originalTypeName.resolveRawType()
+                    .simpleNames()
+                    .joinToString(".")})")
+            .build())
+        .addFunction(FunSpec.builder("fromJson")
+            .addModifiers(OVERRIDE)
+            .addParameter(reader)
+            .returns(originalTypeName)
+            .apply {
+              localProperties.values.forEach {
+                addCode("%L", it.first)
+                it.second?.let {
+                  addCode("%L", it)
+                }
+              }
+            }
+            .addStatement("%N.beginObject()", reader)
+            .beginControlFlow("while (%N.hasNext())", reader)
+            .beginControlFlow("when (%N.selectName(%N))", reader, optionsProperty)
+            .apply {
+              optionsByIndex.map { (index, entry) -> index to entry.value }
+                  .forEach { (index, prop) ->
+                    val specs = localProperties[prop]!!
+                    val spec = specs.first
+                    val setterSpec = specs.second
+                    if (setterSpec != null) {
+                      beginControlFlow("%L -> ", index)
+                      addStatement("%N = %N.fromJson(%N)",
+                          spec,
+                          adapterProperties[prop.typeName to prop.jsonQualifiers]!!,
+                          reader)
+                      addStatement("%N = true", setterSpec)
+                      endControlFlow()
+                    } else {
+                      addStatement("%L -> %N = %N.fromJson(%N)",
+                          index,
+                          spec,
+                          adapterProperties[prop.typeName to prop.jsonQualifiers]!!,
+                          reader)
+                    }
+                  }
+            }
+            .beginControlFlow("-1 ->")
+            .addCode("// Unknown name, skip it.\n")
```

Comment:
`addComment()` can add `\\` and `\n` for you

Reply:
Ahh good catch! 7525f89